src/devices/src/virtio/block/device.rs

```diff
@@ -95,10 +95,12 @@ pub struct Block {
     interrupt_status: Arc<AtomicUsize>,
     interrupt_evt: EventFd,
     pub(crate) queue_evts: [EventFd; 1],
-
     pub(crate) device_state: DeviceState,
 
     // Implementation specific fields.
+    id: String,
+    partuuid: Option<String>,
+    root_device: bool,
```

Comment:
Virtio block doesn't care about partuuid, if the block device is root or not and about the id. While I understand simplification of code as an argument for having these here, I don't think we should go this far.

How about we create a wrapper over `VirtioBlock` in which we extend the Block device struct with the fields that are Firecracker specific?

Reply:
Would be against doing that incrementally? All of our devices have Firecracker-specific logic in them right now.
Decoupling them from that logic needs careful thinking and design so it works uniformly for all devices.

I would prefer not to burden the current step/PR with that responsibility. I don't think having some plain-old-data in the device (like partuuid and id and root bool) is dangerous in any way (like we already have `mmds` in net for example).

Later on, when we get to decoupling the devices from firecracker-specific impl/logic we can think of the proper solution.

What do you think?

Reply:
All virtio devices unfortunately have other state in them (mmds in net, rate limiter in net & block). But I would be worry to add even more non-virtio specific logic to it because we would need to remove it later on. This solution of merging `BlockDeviceConfig` with `VirtioDevice` is only a temporary one. For the final solution we would need to create that separate VirtioDevice wrapper for Firecracker anyway, so I was thinking we can do this now.

With this change, we are introducing API logic to the Virtio Device, which was previously in the VMM. So if we go down this road, after fixing https://github.com/firecracker-microvm/firecracker/issues/1713, we will need to create another issue for "decouple virtio devices from Firecracker specific RPC interface". The fact that we have an id in Virtio Block is RPC specific.

What are your thoughts on this?

Reply:
I disagree that any of that data is rpc or api specific. The id in the device is used in mmio hashtable to be able to find a device, not needed for rpc or api in any way. The `partuuid` is needed for kernel cmdline which is the same, so is the `root_flag`.

The current PR(s) move things in the right direction on multiple fronts with this disadvantage correctly called out by you that devices now have to hold a `String` or a `bool` or some plain data. It is worth mentioning that they don't take dependency on any struct or crate because of this. It's just primitives needed by upper layers.

I agree with you that _at some point_ we will decouple the strict _generic_ block/net/vsock device from the firecracker-impl-specific parts but as of now we don't know these and can't do them. Maybe this ID will be used in rust-vmm mmio-indexing as well - you have to use _something_. Even the device in rust-vmm will need to modify the cmdline if it's not PCI so it might need partuuid or root-block info. The truth is we don't know yet.

As long as we don't know what the most future-proof solution is (we need to define that in rust-vmm), I say we go with the cleanest solution for Firecracker. What do you think?

Reply:
I don't think this is related at all to rust-vmm. IMHO, Virtio Devices should do only Virtio Device specific things. We can take the shortest path for now, but I think we are moving the problem of separation of concerns to lower abstraction levels by adding this fields here.

I dismissed my review.

Reply:
That’s right, we should solve all of these in a unified way alongside the rest tracked in https://github.com/firecracker-microvm/firecracker/issues/1408

I’ll try to do that in the near future when I get some free cycles.